Export/process SnpLz cell properties in a logical order

JanKanis opened the pull request:
This ensures that exported data has its colums written in a predictable and logical order

Fixes NAA-81

JanKanis committed:
Export/process SnpLz cell properties in a logical order

This ensures that exported data has its colums written in a predictable and logical order

Fixes NAA-81

JanKanis commented:
This PR has a (trivial) conflict with the long alleles PR

cataphract commented:
I thinks this is a bad strategy. It makes the correctness of the client code depend on (untested!) implementation details of `SnpLzAllDataProjection`. Nowhere is it guaranteed that the map returned by an `AllDataProjection` has any meaningful order.

At the very least, it should be made clear in the documentation of `SnpLzAllDataProjection` that it provides stronger guarantees that those of `AllDataProjection`, preferably by changing the return type of  `getData/RowProperties` to something that guarantees ordering, say `OrderedMap` or `LinkedHashMap` and use an appropriate implementation (note that `ImmutableMap` is _not_ documented to guarantee this!). And of course, **test** this ordering.

Maybe it is, however, more appropriate to enforce this ordering on the client.

JanKanis committed:
AllDataProjection.{data,row}Properties guarantees a stable and sane iteration order

TabSeparatedExporter uses the order from dataProperties for its columns, and probably other exporters/grid view etc do as well. Until now
AllDataProjection just provided a Map for dataProperties and rowProperties without making any iteration guarantees, but the implementation
already used maps with stable iteration in all cases. This commit formalizes that by specifying it in the AllDataProjection and
MultiValueProjection interfaces. Unfortunately there is no widely used interface that signifies predictable iteration order maps.

This commit adds a test for SnpLzAllDataProjection, which is currently the only place where we depend on a certain order.

This replaces the implementation of data/rowProperties for all high dim datatypes by ImmutableMap-s, which are a bit simpler and memory
efficient compared to .asImmutable() views. I have changed all the related core db code to use ImmutableMap instead of Map, to make sure we
keep the stable iteration order. It would be preferable to use a ordered map interface, but there is none that is applicable here. Using a
specific ordered map implementation in core db and noting it in the api docs for core api seems to be the next best thing.

JanKanis commented:
`ImmutableMap` does guarantee iteration order ([link](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/ImmutableMap.html)). 

There does not seem to be any widely used interface to mark that a Map has a reliable/usable iteration order, so I just noted it in the api docs for `AllDataProjection`/`MultiValueProjection`, and replaced the core-db internal code to use `ImmutableMap` everywhere.

AFAIK all clients that at some point show data from core api to users (exports, rest-api, grid view) use the order from the `dataProperties` iteration, and thus benefit from a stable order in core api. All high dim datatypes already used a stable ordering implementation except for `snp_lz`.
